src/AnalyticsEventReporter.ts

```diff
@@ -1,38 +1,78 @@
-import { AnalyticsEventService } from "./AnalyticsEventService"
-import { AnalyticsConfig } from "./AnalyticsConfig";
-import { EventPayload } from "./EventPayload";
-import { EventAPIResponse } from "./EventAPIResponse";
-import { merge } from "./merge";
+import { AnalyticsEventService } from './AnalyticsEventService';
+import { AnalyticsConfig } from './AnalyticsConfig';
+import { EventPayload, PartialPayload } from './EventPayload';
+import { EventAPIResponse } from './EventAPIResponse';
+import { getOrSetupSessionId } from './setupSessionId';
+import { name, version } from '../package.json';
+import { postWithBeacon, postWithFetch, useBeacon } from './post';
+import merge from './merge';
+import { setupRequestUrl } from './setupRequestUrl';
 
 /** Represents an reporter is responsible for reporting analytics events. */
 export class AnalyticsEventReporter implements AnalyticsEventService {
-
     private config: AnalyticsConfig;
     private payload?: EventPayload;
-
     /**
      * @param config - necessary analytics config: Must provide one and only
      * one of API Key or Bearer Token.
-     * 
+     *
      * @param payload - (optional) desired event values to report
      */
     constructor(config: AnalyticsConfig, payload?: EventPayload) {
-        const apiKeyIsSet = config.key !== undefined;
-        const bearerTokenIsSet = config.bearer !== undefined;
-        const configIsValid = (apiKeyIsSet || bearerTokenIsSet) && !(apiKeyIsSet && bearerTokenIsSet);
-        if (!configIsValid) {
-            throw new Error("Provide one and only one of API Key or Bearer Token.")
-        }
-        this.config = config;
-        this.payload = payload;
+      const apiKeyIsSet = config.key !== undefined;
+      const bearerTokenIsSet = config.bearer !== undefined;
+      const configIsValid = (apiKeyIsSet || bearerTokenIsSet) && !(apiKeyIsSet && bearerTokenIsSet);
+      if (!configIsValid) {
+        throw new Error('Provide one and only one of API Key or Bearer Token.');
+      }
+      this.config = config;
+      this.payload = payload;
     }
 
     with(payload: EventPayload): AnalyticsEventService {
-        const currentPayload = this.payload === undefined ? payload : merge(this.payload, payload);
-        return new AnalyticsEventReporter(this.config, currentPayload);
+      const currentPayload = this.payload === undefined ? payload : merge(this.payload, payload);
+      return new AnalyticsEventReporter(this.config, currentPayload);
     }
 
-    report(payload: object): Promise<EventAPIResponse> | boolean {
-        return false;
+    public async report(newPayload: PartialPayload): Promise<boolean | EventAPIResponse> {
+      const finalPayload = this.payload === undefined ? newPayload : merge(this.payload, newPayload);
```

Comment:
Thoughts on using just `merge(this.payload, newPayload)`? Since `merge()` now just returns one of the arguments if either is empty, we can just edit the logic a bit further to also have undefined/null be equivalent to an empty object.

Reply:
Discussed offline. We will handle the case of newPayload or original being empty in the merge function itself. In report, we will pass an empty record if its null or undefined.